File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formmanagement/PreviouslyDownloadedOnlyTest.java
package org.odk.collect.android.feature.formmanagement;

import android.Manifest;

import androidx.test.rule.GrantPermissionRule;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.odk.collect.android.R;
import org.odk.collect.android.support.CollectTestRule;
import org.odk.collect.android.support.CopyFormRule;
import org.odk.collect.android.support.NotificationDrawerRule;
import org.odk.collect.android.support.TestDependencies;
import org.odk.collect.android.support.TestRuleChain;
import org.odk.collect.android.support.pages.GetBlankFormPage;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.assertion.ViewAssertions.doesNotExist;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.empty;
import static org.hamcrest.Matchers.is;

public class PreviouslyDownloadedOnlyTest {

    public TestDependencies testDependencies = new TestDependencies();
    public NotificationDrawerRule notificationDrawer = new NotificationDrawerRule();
    public CollectTestRule rule = new CollectTestRule();

    @Rule
    public RuleChain chain = TestRuleChain.chain(testDependencies)
            .around(GrantPermissionRule.grant(Manifest.permission.GET_ACCOUNTS))
            .around(notificationDrawer)
            .around(new CopyFormRule("one-question.xml"))
            .around(new CopyFormRule("two-question.xml"))
            .around(rule);

    @Test
    public void whenPreviouslyDownloadedOnlyEnabled_notifiesOnFormUpdates_automaticallyAndRepeatedly() {
        rule.startAtMainMenu()
                .setServer(testDependencies.server.getURL())
                .enablePreviouslyDownloadedOnlyUpdates();

        testDependencies.server.addForm("One Question Updated", "one_question", "2", "one-question-updated.xml");
        testDependencies.scheduler.runDeferredTasks();
        notificationDrawer.open()
                .assertAndDismissNotification("ODK Collect", "Form updates available");

        testDependencies.server.addForm("Two Question Updated", "two_question", "1", "two-question-updated.xml");
        testDependencies.scheduler.runDeferredTasks();
        notificationDrawer.open()
                .assertAndDismissNotification("ODK Collect", "Form updates available");
    }

    @Test
    public void whenPreviouslyDownloadedOnlyEnabled_clickingOnNotification_navigatesToGetBlankForm() {
        rule.startAtMainMenu()
                .setServer(testDependencies.server.getURL())
                .enablePreviouslyDownloadedOnlyUpdates();

        testDependencies.server.addForm("One Question Updated", "one_question", "2", "one-question-updated.xml");
        testDependencies.scheduler.runDeferredTasks();

        notificationDrawer.open()
                .clickNotification("Collect", "Form updates available", "Get Blank Form", new GetBlankFormPage())
                .assertText(R.string.newer_version_of_a_form_info)
                .assertOnPage();
    }

    @Test
    public void whenPreviouslyDownloadedOnlyEnabled_getBlankFormsIsAvailable() {
        rule.startAtMainMenu()
                .enablePreviouslyDownloadedOnlyUpdates()
                .assertText(R.string.get_forms);
    }

    @Test
    public void whenPreviouslyDownloadedOnlyEnabled_fillBlankFormRefreshButtonIsGone() {
        rule.startAtMainMenu()
                .enablePreviouslyDownloadedOnlyUpdates()
                .clickFillBlankForm();

        onView(withId(R.id.menu_refresh)).check(doesNotExist());
    }

    @Test
    public void whenPreviouslyDownloadedOnlyDisabled_stopsCheckingForUpdates() {
        rule.startAtMainMenu()
                .setServer(testDependencies.server.getURL())
                .enablePreviouslyDownloadedOnlyUpdates()
                .enableManualUpdates();

        assertThat(testDependencies.scheduler.getDeferredTasks(), is(empty()));
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/feature/formmanagement/PreviouslyDownloadedOnlyTest.kt
package org.odk.collect.android.feature.formmanagement

import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.assertion.ViewAssertions.doesNotExist
import androidx.test.espresso.matcher.ViewMatchers.withId
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.odk.collect.android.R
import org.odk.collect.android.support.TestDependencies
import org.odk.collect.android.support.pages.ErrorPage
import org.odk.collect.android.support.pages.MainMenuPage
import org.odk.collect.android.support.rules.CollectTestRule
import org.odk.collect.android.support.rules.NotificationDrawerRule
import org.odk.collect.android.support.rules.TestRuleChain

class PreviouslyDownloadedOnlyTest {

    private val testDependencies = TestDependencies()
    private val notificationDrawerRule = NotificationDrawerRule()
    private val rule = CollectTestRule(useDemoProject = false)

    @get:Rule
    var ruleChain: RuleChain = TestRuleChain.chain(testDependencies)
        .around(notificationDrawerRule)
        .around(rule)

    @Test
    fun whenPreviouslyDownloadedOnlyEnabled_notifiesOnFormUpdates_automaticallyAndRepeatedly() {
        rule.withProject(testDependencies.server, "one-question.xml", "two-question.xml")
            .setServer(testDependencies.server.url)
            .enablePreviouslyDownloadedOnlyUpdates()

        testDependencies.server.addForm(
            "One Question Updated",
            "one_question",
            "2",
            "one-question-updated.xml"
        )
        testDependencies.scheduler.runDeferredTasks()

        notificationDrawerRule.open()
            .assertNotification(
                "ODK Collect",
                "Form updates available",
                testDependencies.server.hostName
            )
            .clearAll()

        testDependencies.server.addForm(
            "Two Question Updated",
            "two_question",
            "1",
            "two-question-updated.xml"
        )

        testDependencies.scheduler.runDeferredTasks()

        notificationDrawerRule.open()
            .assertNotification(
                "ODK Collect",
                "Form updates available",
                testDependencies.server.hostName
            )
            .clickNotification(
                "ODK Collect",
                "Form updates available",
                MainMenuPage()
            )
    }

    @Test
    fun whenPreviouslyDownloadedOnlyEnabledWithAutomaticDownload_checkingAutoDownload_downloadsUpdatedForms_andDisplaysNotification() {
        rule.withProject(testDependencies.server, "one-question.xml")
            .enablePreviouslyDownloadedOnlyUpdatesWithAutomaticDownload()

        testDependencies.server.addForm(
            "One Question Updated",
            "one_question",
            "2",
            "one-question-updated.xml"
        )

        testDependencies.scheduler.runDeferredTasks()

        notificationDrawerRule.open()
            .assertNotification(
                "ODK Collect",
                "Forms download succeeded",
                "All downloads succeeded!"
            )
            .clickNotification(
                "ODK Collect",
                "Forms download succeeded",
                MainMenuPage()
            ).clickFillBlankForm()
            .assertText("One Question Updated")
    }

    @Test
    fun whenPreviouslyDownloadedOnlyEnabledWithAutomaticDownload_checkingAutoDownload_downloadsUpdatedForms_andDisplaysNotificationWhenFails() {
        testDependencies.server.errorOnFetchingForms()

        val mainMenuPage =
            rule.withProject(testDependencies.server, "one-question.xml")
                .enablePreviouslyDownloadedOnlyUpdatesWithAutomaticDownload()

        testDependencies.server.addForm(
            "One Question Updated",
            "one_question",
            "2",
            "one-question-updated.xml"
        )

        testDependencies.scheduler.runDeferredTasks()

        mainMenuPage.clickFillBlankForm()
            .assertFormDoesNotExist("One Question Updated")

        notificationDrawerRule.open()
            .assertNotification(
                "ODK Collect",
                "Forms download failed",
                "1 of 1 downloads failed!"
            )
            .clickAction(
                "ODK Collect",
                "Show details",
                ErrorPage()
            )
    }

    @Test
    fun whenPreviouslyDownloadedOnlyEnabled_getBlankFormsIsAvailable() {
        rule.withProject(testDependencies.server.url)
            .enablePreviouslyDownloadedOnlyUpdates()
            .assertText(org.odk.collect.strings.R.string.get_forms)
    }

    @Test
    fun whenPreviouslyDownloadedOnlyEnabled_fillBlankFormRefreshButtonIsGone() {
        rule.withProject(testDependencies.server.url)
            .enablePreviouslyDownloadedOnlyUpdates()
            .clickFillBlankForm()
        onView(withId(R.id.menu_refresh)).check(doesNotExist())
    }

    @Test
    fun whenPreviouslyDownloadedOnlyDisabled_stopsCheckingForUpdates() {
        rule.withProject(testDependencies.server.url)
            .enablePreviouslyDownloadedOnlyUpdates()
            .enableManualUpdates()

        assertThat(testDependencies.scheduler.deferredTasks, equalTo(emptyList()))
    }
}
